### PR TITLE
 CDAP-21265 : Add support for PKCE

### PR DESCRIPTION
**Pull request by sahusanket**
### **Title:** 
feat: Add PKCE support for OAuth providers

### **Description:**
This PR introduces Proof Key for Code Exchange (PKCE) support for OAuth providers in CDAP data pipelines.

**Key Changes:**
* **PKCE Flow Support:** Updates the `OAuthStore` schema to include `authType` (expanding the table spec to 6 columns) to properly identify and route PKCE-enabled providers. 
* **Secure State & TTLs:** Adds endpoints to generate the `code_challenge` and securely store the `code_verifier` with a configurable time-to-live (`security.auth.oauth.pkce.code.verifier.ttl.sec`) 

### **TESTING:**

Tested the following cases for a secure store as GCP SECRET MANAGER and the default CDAP secure store. 

### 1. PKCE Flow (AuthType = PKCE)
Tests specifically targeting the Proof Key for Code Exchange logic and the new 10-minute Secure Store TTL.

| Phase | Test Scenario | Action | Expected Result |
| :--- | :--- | :--- | :--- |
| **Config** | Provider Creation | Call `PutOAuthProvider` with `authType = PKCE`. | Provider is saved successfully with PKCE enabled. |
| **Auth URL** | URL Generation | Generate Auth URL for the PKCE provider. | URL contains `code_challenge` and `code_challenge_method=S256`. A `code_verifier` is generated internally. |
| **Storage** | Verifier Persistence | Check Secure Store after Auth URL generation. | The `code_verifier` is saved under the system namespace using the `state` as the key, with a 10-minute TTL. |
| **256 SHA check** | `code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))` | Take the `code_challenge` from step 2, and `code_verifier` from step 3 and Verify. | It should match perfectly. |
| **Callback** | Valid Exchange Payload | Trigger the OAuth callback (mocking the IdP). | The POST body to `tokenRefreshURL` contains `grant_type=authorization_code`, `code`, and the retrieved `code_verifier`. |
| **Storage** | Missing Verifier | Check Secure Store after successful callback. | The `code_verifier` should be deleted. |
| **TTL** | Expiration (Timeout) | Trigger Auth URL generation, wait > 10 minutes, then attempt Token Exchange callback. | Secure Store throws NotFound/null. Token exchange fails gracefully. |
| **TTL** | Success (Within window) | Trigger Auth URL generation, attempt callback within 10 minutes. | The `code_verifier` is retrieved successfully and token exchange proceeds. |

---

### 2. Standard Flow (AuthType = STANDARD or default)
Tests to ensure existing integrations without PKCE continue to work perfectly.

| Phase | Test Scenario | Action | Expected Result |
| :--- | :--- | :--- | :--- |
| **Config** | Provider Creation | Call `PutOAuthProvider` with `authType = STANDARD` (or missing). | Provider is saved successfully without PKCE flags. |
| **Auth URL** | URL Generation | Generate Auth URL for the `STANDARD` provider. | URL does not contain `code_challenge` or `code_challenge_method`. |
| **Storage** | No Verifier Persistence | Check Secure Store after Auth URL generation. | Secure Store is not hit; no `code_verifier` is stored. |
| **Callback** | Valid Exchange Payload | Trigger the OAuth callback (mocking the IdP). | The POST body to `tokenRefreshURL` contains `grant_type=authorization_code` and `code`, but no `code_verifier`. |

### 3. Parsing of Config from CCONF / cdap-default. 

- Tested with default value of 900 seconds 
- Tested with manual edit of cdap master to 300 seconds

> **Note:** 
> PKCE support is currently introduced for internal use. The time-to-live (TTL) auto-cleanup for `code_verifier`s relies on a Secure Store implementation that natively supports TTL (such as GCP Secret Manager). 
> 
> If the default Secure Store implementation is used, unused `code_verifier`s will not be automatically cleaned up upon expiration. However, this is entirely harmless from a security perspective, and the resulting storage footprint is negligible.
> Follow up Jira : https://cdap.atlassian.net/browse/CDAP-21266

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthProvider.java [hunk 1]
```diff
@@ -74,6 +79,15 @@ public String getUserAgent() {
     return userAgent;
   }
 
+  public enum AuthType {
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthProvider.java]: Please add comments to explain what these Auth Types are.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java [hunk 2]
```diff
@@ -333,6 +345,32 @@ private static List<Field<?>> getRow(OAuthProvider oauthProvider) {
             CREDENTIAL_ENCODING_STRATEGY_COL,
             oauthProvider.getCredentialEncodingStrategy().toString()));
     fields.add(Fields.stringField(USER_AGENT_COL, oauthProvider.getUserAgent()));
+    fields.add(Fields.stringField(AUTH_TYPE_COL, oauthProvider.getAuthType().toString()));
     return fields;
   }
+
+  public void storePKCECodeVerifier(String provider, String state, String codeVerifier) throws Exception {
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java]: supernit: use `write` or `put`, instead of `store` to keep the naming consistent with other methods.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java [hunk 3]
```diff
@@ -333,6 +345,32 @@ private static List<Field<?>> getRow(OAuthProvider oauthProvider) {
...
+    String key = getPKCEKey(provider, state);
+    secureStoreManager.put(NamespaceId.SYSTEM.getNamespace(), key, codeVerifier,
+        "PKCE Code Verifier", Collections.emptyMap(), codeVerifierTTL);
+  }
+
+  @javax.annotation.Nullable
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java]: Can be imported?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java [hunk 4]
```diff
@@ -333,6 +345,32 @@ private static List<Field<?>> getRow(OAuthProvider oauthProvider) {
...
+  public String getPKCECodeVerifier(String provider, String state) {
+    String key = getPKCEKey(provider, state);
+    try {
+      return new String(secureStore.getData(NamespaceId.SYSTEM.getNamespace(), key), java.nio.charset.StandardCharsets.UTF_8);
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java]: can be imported?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java [hunk 5]
```diff
@@ -333,6 +345,32 @@ private static List<Field<?>> getRow(OAuthProvider oauthProvider) {
...
+    } catch (Exception e) {
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java:365]: Catch specific exception?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java [hunk 6]
```diff
@@ -333,6 +345,32 @@ private static List<Field<?>> getRow(OAuthProvider oauthProvider) {
...
+      return null;
+    }
+  }
+
+  public void deletePKCECodeVerifier(String provider, String state) {
+    String key = getPKCEKey(provider, state);
+    try {
+      secureStoreManager.delete(NamespaceId.SYSTEM.getNamespace(), key);
+    } catch (Exception e) {
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java:376]: catch specific exception

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java [hunk 7]
```diff
@@ -333,6 +345,32 @@ private static List<Field<?>> getRow(OAuthProvider oauthProvider) {
...
+      // Ignore if not found
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java]: Check if the error was not found.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java [hunk 8]
```diff
@@ -333,6 +345,32 @@ private static List<Field<?>> getRow(OAuthProvider oauthProvider) {
...
+        "PKCE Code Verifier", Collections.emptyMap(), codeVerifierTTL);
+  }
+
+  @javax.annotation.Nullable
+  public String getPKCECodeVerifier(String provider, String state) {
+    String key = getPKCEKey(provider, state);
+    try {
+      return new String(secureStore.getData(NamespaceId.SYSTEM.getNamespace(), key), java.nio.charset.StandardCharsets.UTF_8);
+    } catch (Exception e) {
+      return null;
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java]: Just curious, why is null returned here?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthProvider.java (hunk 1 above)
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthProvider.java]: Can be moved to a separate file as the scope of this enum is not limited to this class. It is referenced in other java packages.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java [hunk 9]
```diff
@@ -70,12 +75,25 @@ public class OAuthHandler extends AbstractSystemHttpServiceHandler {
...
+  private static final String PREF_PKCE_CODE_VERIFIER_TTL = "oauth.pkce.code.verifier.ttl.sec";
+
   private OAuthStore oauthStore;
 
   @Override
   public void initialize(SystemHttpServiceContext context) throws Exception {
+    Map<String, String> prefs = null;
+    try {
+      prefs = context.getPreferencesForNamespace(NamespaceId.SYSTEM.getNamespace(), true);
+    } catch (Exception e) {
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java]: catch specific exception

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java [hunk 10]
```diff
@@ -70,12 +75,25 @@ public class OAuthHandler extends AbstractSystemHttpServiceHandler {
...
+      LOG.warn("Failed to load preferences for OAuth RTR configuration. Using default values.", e);
+    }
+
     super.initialize(context);
-    this.oauthStore = new OAuthStore(context, context, context.getAdmin());
+
+    this.oauthStore = new OAuthStore(context, context, context.getAdmin(),
+        getPrefOrDefault(prefs, PREF_PKCE_CODE_VERIFIER_TTL, 900));
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java]: define const for default value for better visibility

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java [hunk 11]
```diff
@@ -103,6 +121,21 @@ public void getAuthURL(HttpServiceRequest request, HttpServiceResponder responde
...
+
+      if (oauthProvider.getAuthType() == OAuthProvider.AuthType.PKCE) {
+        String state = UUID.randomUUID().toString();
+        String codeVerifier = generateCodeVerifier();
+        String codeChallenge = generateCodeChallenge(codeVerifier);
+
+        try {
+          oauthStore.storePKCECodeVerifier(provider, state, codeVerifier);
+        } catch (Exception e) {
+          throw new OAuthServiceException(java.net.HttpURLConnection.HTTP_INTERNAL_ERROR, "Failed to store PKCE code verifier", e);
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java]: Can be imported?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/OAuthServiceTest.java [hunk 12]
```diff
@@ -50,7 +50,7 @@ public void testCreateProvider() throws IOException {
             clientId,
             clientSecret,
             OAuthProvider.CredentialEncodingStrategy.FORM_BODY,
-            null);
+            null, null);
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/OAuthServiceTest.java:53]: Add some test with not null

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java [hunk 13]
```diff
@@ -590,4 +663,15 @@ void respond(HttpServiceResponder responder) {
       }
     }
   }
+
+  private long getPrefOrDefault(java.util.Map<String, String> prefs, String key, long defaultValue) {
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java]: Can be imported?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java [hunk 14]
```diff
@@ -70,12 +75,25 @@ public class OAuthHandler extends AbstractSystemHttpServiceHandler {
...
 
+  // The following settings can be overridden via CDAP System Preferences
+  // Time To Live in seconds for PCKE based code verifier in Secure Store.
+  private static final String PREF_PKCE_CODE_VERIFIER_TTL = "oauth.pkce.code.verifier.ttl.sec";
+
   private OAuthStore oauthStore;
 
   @Override
   public void initialize(SystemHttpServiceContext context) throws Exception {
+    Map<String, String> prefs = null;
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java]: Initialise empty map, so that null checks are not needed at multiple places.

**Review by vsethi09 — CHANGES_REQUESTED** [review]

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java (hunk 5 above)
Comment (sahusanket) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java:365]: Done
In interface > Exception : `if the specified namespace or name does not exist`

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java (hunk 7 above)
Comment (sahusanket) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java]: The `NotFoundException` and `SecureKeyNotFoundException` both are a part of cdap-common which not a dependency in this module.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java (hunk 8 above)
Comment (sahusanket) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java]: Was done to handle it at OauthHandler level to have a better error message for the user but it's not clean .
Removed it.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java (hunk 9 above)
Comment (sahusanket) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java]: Unreadability of Preferences should not stop the start up of Studio service. 
But yes, added catch for the mentioned / known exceptions.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java (hunk 5 above)
Comment (vsethi09) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java:365]: nit: Can be simplified as

```java
catch (IOException | Exception e) {
    throw new OAuthStoreException("Failed to read PKCE code verifier from secure storage", e);
}
```

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java (hunk 7 above)
Comment (vsethi09) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java]: Is it not safe to assume that any exception caught here would be of type `NotFoundException` or `SecureKeyNotFoundException`? There could be other kind of `Exceptions` here, like `RuntimeException`.

Is it safe to ignore the exception and not propagate it?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java [hunk 15]
```diff
@@ -468,18 +536,37 @@ private HttpRequest.Builder buildHttpRequest(String body,
...
+  private String generateCodeVerifier() {
+    SecureRandom sr = new SecureRandom();
+    byte[] codeVerifierBytes = new byte[32];
+    sr.nextBytes(codeVerifierBytes);
+    return Base64.getUrlEncoder().withoutPadding().encodeToString(codeVerifierBytes);
+  }
+
+  private String generateCodeChallenge(String codeVerifier) throws OAuthServiceException {
+    try {
+      MessageDigest md = java.security.MessageDigest.getInstance("SHA-256");
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java]: Can be imported?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java [hunk 16]
```diff
@@ -468,18 +524,35 @@ private HttpRequest.Builder buildHttpRequest(String body,
     return requestBuilder;
   }
 
+  private String generateCodeVerifier() {
+    SecureRandom sr = new SecureRandom();
+    byte[] codeVerifierBytes = new byte[32];
+    sr.nextBytes(codeVerifierBytes);
+    return Base64.getUrlEncoder().withoutPadding().encodeToString(codeVerifierBytes);
+  }
```
[the remark this answers is not in the document]
Comment (vsethi09) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java:537]: Please fix.

**Review by vsethi09 — CHANGES_REQUESTED** [review]

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java (hunk 5 above)
Comment (sahusanket) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java:365]: Actually we cannot : there's compilation issue : `Types in multi-catch must be disjoint: 'java. io. IOException' is a subclass of 'java. lang. Exception'`

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java (hunk 7 above)
Comment (sahusanket) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java]: For delete : 
Yes , NotFoundException or SecureKeyNotFoundException is Safe to ignore here. 

For any read operation error, we are catching IOexception. 

But for any other exception yes we should catch .. 

There is a UNCLEAN solution like : 
```
catch (Exception e) {
      if (e.getClass().getName().endsWith("NotFoundException")) {
        // Ignore if already deleted
        return;
      }
      throw new OAuthStoreException("Failed to delete PKCE code verifier from secure storage", e);
    }
```
----

I don't think we should throw exception for NOT FOUND and break the flow. 

----

Moreover , it's a delete op. 
- the `PKCECodeVerifier` anyway should be written with TTL , even if for some runtime it is missed, it will be cleaned up Eventually [ for GCP secret manager ] 

Let me know your thoughts.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java (hunk 5 above)
Comment (vsethi09) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java:365]: I mean, you anyways need to propagate `OAuthStoreException`, so just catch Exception as mentioned below. The inner cause exception has the information about what went wrong.

```java
catch (Exception e) {
    throw new OAuthStoreException("Failed to read PKCE code verifier from secure storage", e);
}
```

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java (hunk 7 above)
Comment (vsethi09) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java]: The comment can be rephrased so:

`Ignore any other exception, like not found.`

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java [hunk 17]
```diff
@@ -69,13 +77,29 @@ public class OAuthHandler extends AbstractSystemHttpServiceHandler {
...
 
   private OAuthStore oauthStore;
 
   @Override
   public void initialize(SystemHttpServiceContext context) throws Exception {
+    Map<String, String> prefs = Collections.emptyMap();
+    try {
+      prefs = context.getPreferencesForNamespace(NamespaceId.SYSTEM.getNamespace(), true);
+    } catch (IOException | IllegalArgumentException | AccessException e) {
+      LOG.warn("Failed to load preferences for OAuth RTR configuration. Using default values.", e);
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java]: Is it safe to ignore the exception and continue with the execution?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java (hunk 7 above)
Comment (vsethi09) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java]: Can be handled similar to other places like so:

```java
catch (Exception e) {
      // If key is not found, then we can safely delete provider. For any other exception, throw it.
      if (!e.getClass().getName().contains("NotFoundException")) {
        throw new OAuthStoreException("Failed to delete client credential from OAuth provider secure storage", e);
      }
```

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java (hunk 17 above)
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java]: Remove the catch block and propagate the exception as it may not be safe to ignore it and continue with the execution.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java [hunk 18]
```diff
@@ -590,4 +678,15 @@ void respond(HttpServiceResponder responder) {
       }
     }
   }
+
+  private long getPrefOrDefault(Map<String, String> prefs, String key, long defaultValue) {
+    if (prefs.containsKey(key)) {
+      try {
+        return Long.parseLong(prefs.get(key));
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java]: Instead of using default. Let the exception propagate. Read this configuration from cConf.

**Review by vsethi09 — CHANGES_REQUESTED** [review]

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/service/http/BasicSystemHttpServiceContext.java [hunk 19]
```diff
@@ -214,4 +214,9 @@ public List<NamespaceSummary> listNamespaces() throws Exception {
   public ContextAccessEnforcer getContextAccessEnforcer() {
     return contextAccessEnforcer;
   }
+
+  @Override
+  public Map<String,String> getConfiguration(String prefix) {
```
Comment (vsethi09) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/service/http/BasicSystemHttpServiceContext.java]: nit: missing space (` `) before `,`

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java [hunk 20]
```diff
@@ -69,13 +76,23 @@ public class OAuthHandler extends AbstractSystemHttpServiceHandler {
...
+  private static final String PREF_PKCE_CODE_VERIFIER_TTL = "pkce.code.verifier.ttl.sec";
 
   private OAuthStore oauthStore;
 
   @Override
   public void initialize(SystemHttpServiceContext context) throws Exception {
     super.initialize(context);
-    this.oauthStore = new OAuthStore(context, context, context.getAdmin());
+    Map<String, String> conf = context.getConfiguration(OAUTH_CONF_PREFIX);
+    this.oauthStore = new OAuthStore(context, context, context.getAdmin(),
```
Comment (vsethi09) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java]: Pass the configuration `Map<String, String>` in the `OAuthStore` constructor, so that you don't have to modify it in the future when more options are needed.

**Review by vsethi09 — APPROVED** [review]

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthProvider.java [hunk 21]
```diff
@@ -33,18 +33,23 @@ public class OAuthProvider {
   @Nullable
   private final String userAgent;
 
+  @Nullable
+  private final AuthType authType;
```
[the remark this answers is not in the document]
Comment (deepika-lalwani123) [reply cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthProvider.java:37]: Please fix.

**Review by deepika-lalwani123 — APPROVED** [review]
